lib/replset.js

```diff
@@ -681,14 +681,14 @@ class ReplSet extends EventEmitter {
             if (
               ismaster.ismaster &&
               ismaster.electionId &&
-              !self.electionId.equals(ismaster.electionId)
+              !ismaster.electionId.equals(self.electionId)
```

Comment:
While this is an actual bug, I think we'd be doing ourselves a disservice by not fixing the actual problems with this method. As far as I read here, there are two issues here:
  - (most importantly) the whole while loop is wrapped in a `try/catch` and indiscriminately delayed and retried on all errors. I consider this a heinous bug, which would have saved you a lot of time during debugging
  - (far less importantly) This is another case where the method can just return `co`, rather than wrapping it in a `Promise` again, further leading to confusion about when and where errors should be caught.

I'm still confused about why the `electionId` isn't there, but we might be missing many other errors here because of the earlier identified bug.

Reply:
a) Great, let's wait until the next PR for those then
b) This might be more of a case of re-throwing anything that's not generated by this module or the driver. The general way of doing this would be to create a base error type for this module (`TopologyManagerError`?) and converting all of our `new Error(...` to use that. Alternatively, and a somewhat more brittle solution, would be to simply check if the thrown error is an instance of the node error types: `AssertionError`, `RangeError`, `ReferenceError`, `SyntaxError`, `TypeError`. I would greatly prefer the first approach but understand its a lot more work!